Adaptive Content Block: Send results to external service when learner submits review question

itsjeyd opened the pull request:
This PR adds logic for notifying Domoscio when a learner submits a review question.

cf. [OC-2152](https://tasks.opencraft.com/browse/OC-2152)

**Implementation**

When a learner submits a review question, ACB needs to send an `EventResult` event for the "knowledge node student" object that links the learner to the review question to Domoscio. The event must include information about the result that the learner obtained (pass/fail).

To make this possible, this PR introduces a new `TRACKING_BACKEND` called `AdaptiveLearningBackend`. Like other existing backends, the `AdaptiveLearningBackend` listens to events that the platform generates when users interact with courses. When it receives a [`problem_check` event](https://edx.readthedocs.io/projects/devdata/en/latest/internal_data_formats/tracking_logs.html#problem-check), it extracts relevant information from it and passes it to `AdaptiveLibraryContentModule`, which takes care of sending an appropriate event to Domoscio.

**Test instructions**

1. Follow instructions on the [epic](https://tasks.opencraft.com/browse/OC-2127) to set up a FUN box.

2. Switch to this branch (`adaptive-content-block`) and install the new block via

    ```sh
    pip install -r requirements/edx/local.txt
    ```

3. Create a new content library and add a few problems to it.

4. Create a new course and add `"adaptive_library_content"` to advanced modules.

5. Add an "Adaptive Content Block" that references the library created previously to a unit (optionally modifying the number of components to display to each student), and publish the unit. Do **not** access the unit in the LMS yet.

6. To be able to work with the Domoscio API, you'll need to specify a few course-level settings: In Studio, navigate to the Advanced Settings page for the course ("Paramètres" > "Paramètres avancés") and replace the default value for the "Adaptive Learning Configuration" setting with the following:

    ```json
    "url": "<see comment on internal ticket>",
    "instance_id": <see comment on internal ticket>,
    "access_token": "<see comment on internal ticket>",
    "api_version": "v1"
    ```

7. In a venv that has the `requests` library installed, create a knowledge node that corresponds to the unit you created earlier (`block_id` needs to be a 32-character hash), and review the results:

    ```python
    import requests, json
    response = requests.post("<url>/v1/instances/<instance_id>/knowledge_nodes", headers={"Authorization": "Token token=<access_token>"}, data={"knowledge_graph_id": <see comment on internal ticket>, "name": "Unit node", "uid": "<block_id of unit>"})
    json.loads(response.content)
    ```

    Then create a knowledge node for each of the problems that belong to the library you created earlier. Note that problems belonging to a content library are assigned unique block IDs for each Adaptive Content Block that references them; you'll need to use these context-specific IDs when creating corresponding knowledge nodes. Block IDs for problems are 20-character hashes.

8. Review the list of events that exist on the Domoscio test instance:

    ```python
    events = requests.get("<url>/v1/instances/<instance_id>/events", headers={"Authorization": "Token token=<access_token>"})
    json.loads(events.content)
    ```

9. Access the unit that contains the ACB in the LMS. It will display all of the child blocks belonging to the ACB. *This is just for testing purposes.* (You won't want to wait until reviews become available :)) I'll remove the [line](https://github.com/open-craft/edx-platform/pull/50/files#diff-5b223f1aa12d0cba135cf45dade27de3R555) that disables the selection logic introduced in #49 before merging this PR.

10. Review the list of events as described above. It should now contain one more event (this is a "Unit viewed" event that gets sent when accessing a unit containing an ACB in the LMS; it should reference the "knowledge node student" object linking the unit you created earlier to the student that you used to access the unit in the LMS).

11. Answer one of the review questions and submit it (click "Vérifier").

12. Review the list of events as described above. It should now contain one more event. This is the result event that notified Domoscio about the fact that you submitted a review question, and what the results were. It should reference the "knowledge node student" object linking the review question you submitted to the current user, and its `"payload"` property should be set to the value corresponding to the result you obtained for the review question (`"0"` if you got the question wrong, `"100"` if you got it right).

**Relevant tests**

Before running any of the commands listed below, remove (or comment out) the [line](https://github.com/open-craft/edx-platform/pull/50/files#diff-5b223f1aa12d0cba135cf45dade27de3R555) that disables the selection logic introduced in #49.

The commands for running relevant tests are:

```sh
paver test_system -t lms/djangoapps/course_blocks/transformers/tests/test_library_content.py --fasttest  # Didn't make changes to this file, but the tests it defines should still be passing
paver test_lib -l common/lib/xmodule/xmodule/tests/test_library_content.py
paver test_lib -l common/lib/xmodule/xmodule/tests/test_util_adaptive_learning.py
paver test_system -t common/djangoapps/track/backends/tests/test_adaptive_learning.py --fasttest
```

**Reviewers**

- [ ] @e-kolpakov / @pomegranited

**Notes**

- The choice of result values that get sent to Domoscio on pass/fail (`"100"` and `"0"`, respectively) is based on information from the Domoscio API documentation. I'm just mentioning this here because I was surprised initially that the values had to be sent as strings.

e-kolpakov commented on common/djangoapps/track/backends/adaptive_learning.py:
```diff
@@ -0,0 +1,83 @@
+    Return True if `event` is of type `problem_check`, else False.
+    """
+    event_type = event.get('event_type')
+    return event_type == 'problem_check'
```
@itsjeyd nit: inline `event_type`

e-kolpakov commented on common/djangoapps/track/backends/adaptive_learning.py:
```diff
@@ -0,0 +1,83 @@
+    module = context.get('module')
+    usage_key_string = module.get('usage_key')
+    usage_key = UsageKey.from_string(usage_key_string)
+    block_id = usage_key.block_id
```
@itsjeyd nit: inline `block_id` and probably `usage_key`

e-kolpakov commented on common/djangoapps/track/backends/adaptive_learning.py:
```diff
@@ -0,0 +1,83 @@
+    Return success information from `event`.
+    """
+    event = event.get('event')
+    success = event.get('success')
```
@itsjeyd and inline `success`

e-kolpakov replied:
@itsjeyd and also there's a lot of this pattern:

    part1 = event.get('part1')
    part2 = event.get('part2')

Maybe it would be advantageous to have something like

    def get_from_event(event, path, default):
        parts = path.split('.')
        for part in parts[:-1]:
            current = event.get(part, {})
        return current.get(parts[-1], default)

And use it like

    get_from_event(event, 'event.success', None)

itsjeyd replied:
@e-kolpakov Thanks for suggesting to introduce a separate method for extracting data from the `event` dictionary. I [implemented it](https://github.com/open-craft/edx-platform/pull/50/files#diff-de65d67fa632ac825614b5bfd8d28bd7R24).

e-kolpakov commented on common/djangoapps/track/backends/adaptive_learning.py:
```diff
@@ -0,0 +1,83 @@
+    Return ID of user that triggered `event`.
+    """
+    context = event.get('context')
+    user_id = context.get('user_id')
```
@itsjeyd and inline `user_id`

e-kolpakov commented on common/djangoapps/track/backends/tests/test_adaptive_learning.py:
```diff
@@ -0,0 +1,154 @@
+        ('other', False)
+    )
+    @ddt.unpack
+    def test__is_problem_check(self, event_type, expected_result):
```
@itsjeyd nononononono, don't test private methods!!11! :)

e-kolpakov commented on common/djangoapps/track/backends/tests/test_adaptive_learning.py:
```diff
@@ -0,0 +1,154 @@
+            mock_course_key = Mock()
+            patched_class.from_string.return_value = mock_course_key
+            mock_modulestore = Mock()
+            mock_course = Mock()
```
@itsjeyd now you see why `modulestore()` is a bad idea? :) Suggestion: make modulestore passed into `AdaptiveLearningBackend` (as a constructor parameter?), and just pass a mocked instance to `AdaptiveLearningBackend` instantiation.

Oh, yes, and this requires refactoring all those methods into static/class/instancemethods and redoing the test suite into testing `AdaptiveLearningBackend`. Which is something you should do anyway :)

e-kolpakov commented on common/djangoapps/track/backends/tests/test_adaptive_learning.py:
```diff
@@ -0,0 +1,154 @@
+            expected_block_id = '8e52e13fc4g696gb8g33'
+            mock_usage_key = Mock()
+            mock_usage_key.block_id = expected_block_id
+            patched_class.from_string.return_value = mock_usage_key
```
@itsjeyd nit: might make sense to extract a helper method `make_mock_usage_key`

e-kolpakov commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -549,6 +549,10 @@ def get_selections_current_user(self, children):
+        # FIXME:
+        # Make it possible to test changes manually by temporarily showing all children.
+        # Will revert this before merging PR.
+        selections += valid_block_keys.values()
```
@itsjeyd Ok, as a reminder - I'm requesting changes to remove it :)

e-kolpakov commented on common/lib/xmodule/xmodule/tests/test_util_adaptive_learning.py:
```diff
@@ -369,17 +398,72 @@ def test_create_read_event(self):
+        }
+        with patch.object(DummyModule, '_make_anonymous_user_id') as patched__make_anonymous_user_id, \
+             patch.object(DummyModule, '_create_event') as patched__create_event:
+            patched__make_anonymous_user_id.return_value = user_id
```
@itsjeyd I really don't like you mocking private methods/functions. This is a sign of dependency being hard-coded in a form of private method. Please either extract them into a proper form (i.e. new class that is injected into the object that uses it) or mock underlying implementation. I'd suggest the former.

e-kolpakov commented on common/lib/xmodule/xmodule/util/adaptive_learning.py:
```diff
@@ -53,136 +53,163 @@ def _adaptive_learning_configuration(self):
+        return self.__request_headers(self._adaptive_learning_configuration)
+
+    @classmethod
+    def _adaptive_learning_url(cls, adaptive_learning_configuration):
```
@itsjeyd +1 ^^

itsjeyd replied:
@pomegranited Turning the mixin into a class that can be instantiated on its own is a great idea, thanks :) I made that change. Just note that the new class no longer gets mixed in to `AdaptiveLibraryContentModule`: @e-kolpakov mentioned [here](https://github.com/open-craft/edx-platform/pull/49#discussion_r93747252) that

> there's an explicit instruction to never override `__init__` as it might break XBlock lifecycle

and that there is a chance that this might affect XModules as well. So instead of keeping the new class as a parent class for `AdaptiveLibraryContentModule` and giving it an `__init__` that takes an `adaptive_learning_configuration` or `course` parameter, I added an [`api_client` property](https://github.com/open-craft/edx-platform/pull/50/files#diff-5b223f1aa12d0cba135cf45dade27de3R401) to `AdaptiveLibraryContentModule` that instantiates the new class with the block's `parent_course`.

e-kolpakov commented on lms/envs/common.py:
```diff
@@ -617,7 +617,10 @@
-    }
+    },
+    'adaptive_learning': {
+        'ENGINE': 'track.backends.adaptive_learning.AdaptiveLearningBackend',
```
@itsjeyd nit: `ENGINE` or `BACKEND`?

itsjeyd replied:
@e-kolpakov It needs to be `ENGINE`, because the [code](https://github.com/open-craft/edx-platform/blob/send-result-events/common/djangoapps/track/tracker.py#L37) that instantiates the different `TRACKING_BACKENDS` [uses that property](https://github.com/open-craft/edx-platform/blob/send-result-events/common/djangoapps/track/tracker.py#L50) to determine the corresponding Python class.

itsjeyd commented:
@pomegranited @e-kolpakov Thanks for reviewing! I addressed all of your comments. Please have another look. I'll make sure to remove the [FIXME](https://github.com/open-craft/edx-platform/pull/50#discussion_r94292339) before merging this PR.

@pomegranited There are a few quality violations left that came from #47. I'll address those in the context of OC-2154, which will follow up on that PR. The remaining violations coming from adaptive learning code have been fixed.

itsjeyd commented on common/djangoapps/track/backends/adaptive_learning.py:
```diff
@@ -0,0 +1,98 @@
+        Extract value of `path` from `event` dictionary and return it.
+
+        If `path` does not exist in `event`, return `default` value instead.
+        """
```
@pomegranited Nice! I replaced the original implementation with this.

itsjeyd commented:
Thanks @pomegranited!

FIXME removed in c4d8cbc32932ba1c5a8381c759c7e642ae6ca2c6.

itsjeyd commented:
@e-kolpakov Just a friendly reminder to have another look at this :) We'll need to merge ASAP.

itsjeyd commented:
Thanks @e-kolpakov! Merging now.
